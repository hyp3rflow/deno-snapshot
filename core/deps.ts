export * as fs from "https://deno.land/std@0.129.0/fs/mod.ts";
export * as path from "https://deno.land/std@0.129.0/path/mod.ts";
export { delay } from "https://deno.land/std@0.129.0/async/delay.ts";
export { assertEquals } from "https://deno.land/std@0.129.0/testing/asserts.ts";
export { ensureFileSync, existsSync } from "https://deno.land/std@0.129.0/fs/mod.ts";

export { default as JSON5 } from 'https://esm.sh/json5@2.2.0';
export { default as micromatch } from 'https://esm.sh/micromatch@4.0.4?bundle';