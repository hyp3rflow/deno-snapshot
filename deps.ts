export { default as isCi } from "https://esm.sh/is-ci?target=deno";
export { default as JSON5 } from "https://esm.sh/json5";
export { assertEquals } from "https://deno.land/std/testing/asserts.ts";
export * as path from "https://deno.land/std/path/mod.ts";
export { ensureFileSync, existsSync } from "https://deno.land/std/fs/mod.ts";
export { delay } from "https://deno.land/std/async/delay.ts";
export { default as micromatch } from "https://esm.sh/micromatch?bundle";
