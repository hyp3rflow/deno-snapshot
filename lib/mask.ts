import { micromatch } from "../deps.ts";

export default function mask(anyType: unknown, masks: string[]): unknown {
  validateMasks(masks);
  let matches = 0;
  const _mask = (
    currType: unknown,
    _masks: string[],
    dotPath = ``,
  ): unknown => {
    const shouldMask = micromatch.isMatch(dotPath, _masks);
    matches += shouldMask ? 1 : 0;
    // _masks.filter((m) => m === dotPath).length;
    if (currType instanceof Array) {
      return shouldMask
        ? `[MASKED array]`
        : currType.map((el, idx) => _mask(el, _masks, `${dotPath}/${idx}`));
    } else if (currType instanceof Object) {
      return shouldMask ? `[MASKED object]` : Object.entries(currType).reduce(
        (acc: Record<string, unknown>, entry) => {
          const [key, val] = entry;
          acc[key] = _mask(val, _masks, `${dotPath}/${key}`);
          return acc;
        },
        {},
      );
    } else {
      return shouldMask ? `[MASKED ${typeof currType}]` : currType;
    }
  };
  const result = _mask(anyType, masks);

  // if no matches are found, likely the masks are an error
  if (matches === 0 && 0 < masks.length) {
    throw new Error(`None of the masks matched. This is likely an error`);
  }

  return result;
}

/** ensure that each mask starts with '/' */
function validateMasks(masks: string[]) {
  const violations: string[] = masks.filter((m: string) => m[0] !== "/");
  if (violations.length) {
    throw new Error(
      `Masks should begin from root '/'. Violations:\n  ${
        violations.join(
          "\n  ",
        )
      }`,
    );
  }
}
