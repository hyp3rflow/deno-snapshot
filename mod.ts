import { assertEquals, ensureFileSync, JSON5, path } from "./deps.ts";
import mask from "./lib/mask.ts";

const trace = false;
const debug = (msg: string) => {
  if (trace) {
    console.log(msg);
  }
};

type SnapshotFn = (actual: unknown, masks?: string[], name?: string) => void;

type ExecMode = "validate" | "update" | "refresh";

interface TestContext {
  name: string;
  assertSnapshot: SnapshotFn;
  execMode: ExecMode;
}

type WrappedTest = (t: TestContext) => void | Promise<void>;
type WrappedTestDefinition = Omit<Deno.TestDefinition, "fn"> & {
  fn: WrappedTest;
};

/**
 * Wraps Deno.test so the test function accepts an argument of TestContext.
 * TestContext provides a test specific "assertSnapshot" function that can be used to
 */
function compose(
  _test = Deno.test,
): (
  name: string | WrappedTestDefinition,
  fn?: (t: TestContext) => void,
) => void | Promise<void> {
  const execMode = snapshotMode();
  debug(`execMode: ${execMode}`);

  const contextMap: Record<string, string> = {};
  const snapshotsHaveUpdates: Record<string, boolean> = {};
  let count = 0;
  const getCount = () => (count += 1);
  const resetCount = () => {
    count = 0;
  };

  function readSnapshotsFromDisk(snapshotFile: string) {
    if (0 === Object.keys(contextMap).length && execMode === "refresh") {
      // First assertSnapshot in a given file and in refresh mode.
      // Empty the assertSnapshot file, clears any unused snapshots.
      // Clears any unused snapshots.
      Deno.writeTextFileSync(snapshotFile, JSON5.stringify({}));
      snapshotsHaveUpdates[snapshotFile] = true;
      return {};
    }
    ensureFileSync(snapshotFile);
    return JSON5.parse(Deno.readTextFileSync(snapshotFile) || `{}`);
  }

  return (
    name: string | WrappedTestDefinition,
    fn: (t: TestContext) => void = () => {},
  ): void | Promise<void> => {
    // create testDefinition - this will be passed to Deno.test()
    const testDefinition = typeof name === "string"
      ? {
        name,
        fn,
      }
      : name;

    const assertSnapshot: SnapshotFn = (actual, masks = [], title) => {
      const snapshotFile = getSnapshotFileName();
      const snapshots = readSnapshotsFromDisk(snapshotFile);
      if (!Object.keys(contextMap).includes(testDefinition.name)) {
        resetCount();
      }

      const c = getCount();
      title = title && title !== testDefinition.name
        ? title
        : c === 1
        ? testDefinition.name
        : `${testDefinition.name}:#${c}`;
      debug(`\n Snapshot: ${title} ${c}`);
      if (Object.keys(contextMap).includes(title)) {
        throw new Error(
          `Duplicate assertSnapshot title - ${title} ${c} ${name}\n`,
        );
      }
      const expected = snapshots[title] ?? null;
      const _actual = mask(actual, masks);
      switch (execMode) {
        case "refresh": {
          // if first attempt, delete
          snapshots[title] = _actual;
          snapshotsHaveUpdates[snapshotFile] = true;
          break;
        }
        case "validate": {
          if (expected) {
            compareSnapshots(_actual, expected, title, );
          } else {
            snapshots[title] = _actual;
            snapshotsHaveUpdates[snapshotFile] = true;
          }
          break;
        }
        case "update": {
          try {
            compareSnapshots(_actual, expected, title, );
          } catch {
            // There was assertSnapshot mismatch, but we are in update mode.
            // Prevents updating the file needlessly, reducing disk IO.
            snapshots[title] = _actual;
            snapshotsHaveUpdates[snapshotFile] = true;
          }
          break;
        }
      }
      // finally, if we have updates to the snapshotFile, write them out.
      // this is done for each assertSnapshot, makes things a bit slower, but
      // is the most consistent way to keep track
      if (snapshotsHaveUpdates[snapshotFile]) {
        Deno.writeTextFileSync(
          snapshotFile,
          JSON5.stringify(snapshots, { space: 2 }),
        );
      }
    };

    const context = {
      name: testDefinition.name,
      assertSnapshot,
      execMode,
    };
    // Adjust for the fn being a WrappedTest instead of a
    const _fn = testDefinition.fn;
    testDefinition.fn = () => _fn(context);
    // it's only when Deno.test decides to execute the tests, is any
    // of the snapshot functionality exercised.
    return _test(testDefinition as unknown as Deno.TestDefinition);
  };
}
export const test = compose();
export default test;

function getSnapshotFileName() {
  const testFile = path.fromFileUrl(Deno.mainModule);
  const parts = path.parse(testFile);
  return `${parts.dir}/${parts.name}.snap`;
}

/**
 * Given an input of Deno.args, determines update mode.
 * Usage: `snapshotMode(Deno.args)`
 */
export function snapshotMode(args: string[] = Deno.args): ExecMode {
  if (args.includes("-u") || args.includes("--update")) {
    return "update";
  }
  if (args.includes("-r") || args.includes("--refresh")) {
    return "refresh";
  }
  return "validate";
}

export function compareSnapshots(
  actual: unknown,
  expected: unknown,
  title: string,
) {
  try {
    // stringify+parse to ensure that "undefined" fields are removed.
    // `expected` has been through this process
    actual = JSON5.parse(JSON5.stringify(actual));
    assertEquals(actual, expected);
  } catch (err) {
    const message = `Snapshot mismatch:\n  ${title}\n  ${
      err.message
        .split("\n")
        .join("\n  ")
    }`;
    throw new Error(message);
  }
}
