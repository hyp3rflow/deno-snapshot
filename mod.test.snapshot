{
  'assertSnapshot Test': {
    a: 1,
    b: 'b',
  },
  'assertSnapshot Test:#2': {
    c: 2,
    d: 'd',
  },
  'a second test': [
    1,
    2,
    3,
  ],
  'a second test:#2': 'some string',
  'a second test:#3': 2,
  'a second test:#4': true,
  'skip test on update': {
    a: 1,
    u: false,
  },
  'test with masks': {
    blah: [
      {
        blue: '[MASKED number]',
        a: 1,
      },
      {
        blue: '[MASKED number]',
        a: 2,
      },
    ],
    a: {
      s: {
        d: {
          f: '[MASKED boolean]',
          g: true,
        },
        d1: 'd1',
      },
      s1: 's1',
    },
    a1: 'a1',
    b: [
      {
        c: '[MASKED number]',
        d: 1,
      },
      {
        c: 2,
        d: 2,
      },
    ],
  },
  'Test with masks - example': {
    a: 'a',
    b: 1,
    c: {
      start: '[MASKED number]',
      randomArray: [
        '[MASKED number]',
        '[MASKED number]',
      ],
    },
  },
}