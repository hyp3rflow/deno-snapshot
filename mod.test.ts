import test from "./mod.ts";

test(`assertSnapshot Test`, ({ assertSnapshot }) => {
  assertSnapshot({ a: 1, b: "b" });
  assertSnapshot({ c: 2, d: "d" });
});

Deno.test('name', (t) => {
  t.step('step 1', () => {
    
  })
})

test(`a second test`, ({ assertSnapshot }) => {
  assertSnapshot([1, 2, 3]);
  assertSnapshot("some string");
  assertSnapshot(2);
  assertSnapshot(true);
});

test({
  name: `skip test on update`,
  fn: ({ assertSnapshot }) => {
    // this test case is skipped with a --update flag.
    assertSnapshot({
      a: 1,
      u: Deno.args.includes("--update"),
    });
  },
  ignore: Deno.args.includes("--update"),
});

test(`test with masks`, ({ assertSnapshot }) => {
  const ot = {
    blah: [
      { blue: 1, a: 1 },
      { blue: 2, a: 2 },
    ],
    a: {
      s: {
        d: {
          f: true,
          g: true,
        },
        d1: "d1",
      },
      s1: "s1",
    },
    a1: "a1",
    b: [
      { c: 1, d: 1 },
      { c: 2, d: 2 },
    ],
  };

  const masks = ["/blah/*/blue", "/a/s/d/f", "/b/0/c"];
  assertSnapshot(ot, masks);
});

test(`Test with masks - example`, ({ assertSnapshot }) => {
  const actual = {
    a: "a",
    b: 1,
    c: {
      start: Date.now(),
      randomArray: [
        Math.random(),
        Math.random(),
      ],
    },
  };
  const masks = ["/c/start", "/c/randomArray/*"];
  assertSnapshot(actual, masks);
});
