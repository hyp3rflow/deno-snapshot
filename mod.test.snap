{
  'assertSnapshot Test': {
    a: 1,
    b: 'b',
  },
  'a second test': [
    1,
    2,
    3,
  ],
  'skip test on update': {
    a: 1,
    u: false,
  },
  'test with masks': {
    blah: [
      {
        blue: '[MASKED number]',
        a: 1,
      },
      {
        blue: '[MASKED number]',
        a: 2,
      },
    ],
    a: {
      s: {
        d: {
          f: '[MASKED boolean]',
          g: true,
        },
        d1: 'd1',
      },
      s1: 's1',
    },
    a1: 'a1',
    b: [
      {
        c: '[MASKED number]',
        d: 1,
      },
      {
        c: 2,
        d: 2,
      },
    ],
  },
  'Test with masks - example': {
    a: 'a',
    b: 1,
    c: {
      start: '[MASKED number]',
      randomArray: [
        '[MASKED number]',
        '[MASKED number]',
      ],
    },
  },
}